src/operator/contrib/allclose_op-inl.h

```diff
@@ -0,0 +1,138 @@
+      .set_default(1e-08)
+      .describe("Absolute tolerance.");
+    DMLC_DECLARE_FIELD(equal_nan)
+      .set_default(true)
+      .describe("Whether to compare NaN’s as equal. If True, NaN’s in a will be considered equal "
+                "to NaN’s in b in the output array.");
+  }
+};
+
+inline bool AllCloseShape(const nnvm::NodeAttrs& attrs,
+                          std::vector<TShape>* in_attrs,
+                          std::vector<TShape>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U) << "Input:[array1, array2]";
+  CHECK_EQ(out_attrs->size(), 1U);
+
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape());
+  return in_attrs->at(0) == in_attrs->at(1);
+}
+
+inline bool AllCloseType(const nnvm::NodeAttrs& attrs,
+                         std::vector<int> *in_attrs,
+                         std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(out_attrs->size(), 1U);
+
+  // The output will be boolean stored as an integer
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kInt32);
+  return (*out_attrs)[0] != -1;
+}
+
+using namespace mshadow_op::isnan_typed;
+
+template<int req>
+struct allclose_forward {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, int *out_data, const DType* in_a, const DType* in_b,
+                                  const float rtol, const float atol, bool equal_nan) {
+      const DType a = in_a[i], b = in_b[i];
+      const bool val = (IsNan(a) || IsNan(b))? equal_nan :
+                       a == b ||
+                       (a > b?  a - b : b - a) <= atol + (b > 0? rtol * b :  (-rtol) * b);
+      KERNEL_ASSIGN(out_data[i], req, val? 1 : 0);
+  }
+};
+
+template<typename xpu>
+size_t GetAdditionalMemorySize(const int num_items);
+
+template<typename xpu>
+void AllCloseAction(mshadow::Stream<xpu> *s,
+                    int *workspaceMemory,
+                    size_t extraStorageBytes,
+                    const TBlob& in_array0,
+                    const TBlob& in_array1,
+                    const std::vector<OpReqType>& req,
+                    const AllCloseParam& param,
+                    int *outPntr);
+
+template<typename xpu>
+void AllClose(const nnvm::NodeAttrs& attrs,
+              const OpContext& ctx,
+              const std::vector<TBlob>& inputs,
+              const std::vector<OpReqType>& req,
+              const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+
+  const TBlob& in_array0 = inputs[0];
+  const TBlob& in_array1 = inputs[1];
+
+  const AllCloseParam& param = nnvm::get<AllCloseParam>(attrs.parsed);
+
+  auto s = ctx.get_stream<xpu>();
+  const int num_items = in_array0.Size();
+
+  // Get length of the additional memory (which is used only by DeviceReduce::Min(...) on gpu)
+  const size_t extraStorageBytes = GetAdditionalMemorySize<xpu>(num_items);
+  const size_t workspace_total_bytes_ = num_items * sizeof(float) + extraStorageBytes;
```

Comment:
why hardcoded float? Can't we use type_flag? will it work if the type of the tensor is not float?

Reply:
Actually, it was a good catch!

We don't need to use `type_flag` or even `float` here. The correct fix would be
```
  const size_t workspace_total_bytes_ = num_items * sizeof(int) + extraStorageBytes;
```

In my initial implementation `out_data`, which is used in 
```
  MSHADOW_XINLINE static void Map(int i, int *out_data, const DType* in_a, const DType* in_b,
```
was a pointer to `float`. But when I was implementing the possibility for  `in_a, in_b` also be **not-a-numbers**, I 've changed the type of pointer to `int *out_data`.

Perhaps, we could use even `char *out_data`, but I am not sure if 
```
  cub::DeviceReduce::Min(workspaceMemory + num_items, extraStorageBytes,
                         workspaceMemory, outPntr, num_items);
```
would allow for `workspaceMemory` be a `char *`